FeedStoreChallenge/CoreDataFeedStore.swift

```diff
@@ -29,14 +29,48 @@ public final class CoreDataFeedStore: FeedStore {
 	}
 
 	public func retrieve(completion: @escaping RetrievalCompletion) {
-		fatalError("Must be implemented")
+		performContext { context in
+			do {
+				if let cache = try ManagedCache.fetch(in: context) {
+					completion(.found(feed: cache.localFeed, timestamp: cache.timestamp))
+				} else {
+					completion(.empty)
+				}
+			} catch {
+				completion(.failure(error))
+			}
+		}
 	}
 
 	public func insert(_ feed: [LocalFeedImage], timestamp: Date, completion: @escaping InsertionCompletion) {
-		fatalError("Must be implemented")
+		performContext { context in
+			do {
+				let newCache = try ManagedCache.newInstance(in: context)
+				newCache.feed = ManagedCache.feed(from: feed, in: context)
+				newCache.timestamp = timestamp
+				try context.save()
+				completion(nil)
+			} catch {
+				context.rollback()
+				completion(error)
+			}
+		}
 	}
 
 	public func deleteCachedFeed(completion: @escaping DeletionCompletion) {
-		fatalError("Must be implemented")
+		performContext { context in
+			do {
+				try ManagedCache.fetch(in: context).map(context.delete)
+				try context.save()
+				completion(nil)
+			} catch {
+				context.rollback()
+				completion(error)
+			}
+		}
+	}
+
+	private func performContext(_ action: @escaping (NSManagedObjectContext) -> Void) {
+		context.perform { [self] in action(context) }
```

Comment:
There's no need to capture `self` because it can cause `CoreDataFeedStore` not being deallocated in some cases.
It can be optimised just to capturing `context` only using capture list.

```suggestion
		context.perform { [context] in action(context) }
```